### PR TITLE
User params

### PR DESCRIPTION
**Pull request by hagen1778**
This PR allows to create `param_groups` - named lists of URL params. If the group is attached to `user` in `params` option, proxy will send those params with every user's request.

The `param_groups` configuration is following:
```
# Optional list of GET params to send with each request
param_groups:
    # Group name, which may be passed into `params` option on the `user` level.
  - name: "cron-job"
    # List of key-value params to send
    params:
      - key: "max_memory_usage"
        value: "40000000000"

      - key: "max_bytes_before_external_group_by"
        value: "20000000000"
```

Caching is still available while using `param_groups`. Proxy will get a hash from the params list and use it as a part of `cache.Key`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -391,6 +391,29 @@ network_groups:
   - name: "reporting-apps"
     networks: ["10.10.10.0/24"]
 
+# Optional list of GET params to send with each request
```
Comment (valyala) [inline README.md]: substitute this with the following text:

    Optional lists of query params to send with each proxied request to ClickHouse.
    These lists may be used for overriding ClickHouse settings on a per-user basis.

File: README.md [hunk 2]
```diff
@@ -420,7 +443,6 @@ server:
     # Certificates are automatically issued and renewed if this section
     # is present.
     # There is no need in cert_file and key_file if this section is present.
-    # Autocert requires application to listen on :80 port for certificate generation
```
Comment (valyala) [inline README.md]: Why this line has been removed?

File: README.md [hunk 3]
```diff
@@ -467,6 +489,11 @@ users:
     # By default responses aren't cached.
     cache: "longterm"
 
+    # ParamGroup name to use
```
Comment (valyala) [inline README.md]: Substitute this with:

    An optional group of params to send to ClickHouse with each proxied request.
    These params may be set in param_groups block.

    By default no additional params are sent to ClickHouse.

File: config/config.go [hunk 4]
```diff
@@ -464,7 +469,13 @@ func (ng *NetworkGroups) UnmarshalYAML(unmarshal func(interface{}) error) error
 	if err := unmarshal((*plain)(ng)); err != nil {
 		return err
 	}
-	return checkOverflow(ng.XXX, "network_groups")
+	if len(ng.Name) == 0 {
```
Comment (valyala) [inline config/config.go]: Probably, this change should be pushed in a separate commit if it is correct, since it is unrelated to the PR.

File: config/config_test.go [hunk 5]
```diff
@@ -125,6 +160,7 @@ func TestLoadConfig(t *testing.T) {
 						MaxExecutionTime:     Duration(time.Minute),
 						DenyHTTPS:            true,
 						NetworksOrGroups:     []string{"office", "1.2.3.0/24"},
+						Params:               "cron-job",
```
Comment (valyala) [inline config/config_test.go]: Either remove Params from this user or add a user without params in order to verify user config without params works as expected.

File: scope.go [hunk 6]
```diff
@@ -382,6 +391,30 @@ func (s *scope) maxQueueTime() time.Duration {
 	return d
 }
 
+type paramsRegistry struct {
+	// key is a hashed concatenation of the params list
+	key uint32
+
+	r map[string]string
```
Comment (valyala) [inline scope.go]: Why not storing `[]config.Param` instead? This way params will be added to proxied requests always in the same order.

File: README.md (hunk 2 above)
Comment (hagen1778) [reply README.md]: It was replaced with config/testdata/full.yml where this line was absent. Will update add it in both places

File: config/config.go (hunk 4 above)
Comment (hagen1778) [reply config/config.go]: True. Should be fixed in a separate PR

File: scope.go (hunk 6 above)
Comment (hagen1778) [reply scope.go]: Indeed, thx for pointing that out!

File: README.md [hunk 7]
```diff
@@ -488,6 +518,8 @@ users:
     to_user: "default"
     allowed_networks: ["office", "1.2.3.0/24"]
 
+    params: "cron-job"
```
Comment (valyala) [inline README.md]: Probably, this line must be removed in order to match `full.yaml`
